internal/examples/agent/agent/agent.go

```diff
@@ -0,0 +1,280 @@
+
+	// Create agent description.
+	agent.agentDescription = &protobufs.AgentDescription{
+		AgentType:    agent.agentType,
+		AgentVersion: agent.agentVersion,
+		AgentAttributes: []*protobufs.KeyValue{
+			{
+				Key: "os.family",
+				Value: &protobufs.AnyValue{
+					Value: &protobufs.AnyValue_StringValue{
+						StringValue: runtime.GOOS,
+					},
+				},
+			},
+			{
+				Key: "host.name",
+				Value: &protobufs.AnyValue{
+					Value: &protobufs.AnyValue_StringValue{
+						StringValue: hostname,
+					},
+				},
+			},
+		},
+	}
+}
+
+func (agent *Agent) loadLocalConfig() {
+	var k = koanf.New(".")
+	k.Load(rawbytes.Provider([]byte(localConfig)), yaml.Parser())
+
+	effectiveConfigBytes, err := k.Marshal(yaml.Parser())
+	if err != nil {
+		panic(err)
+	}
+
+	agent.effectiveConfig = string(effectiveConfigBytes)
+	hash := sha256.Sum256(effectiveConfigBytes)
+	agent.effectiveConfigHash = hash[:]
+}
+
+func (agent *Agent) composeEffectiveConfig() *protobufs.EffectiveConfig {
+	return &protobufs.EffectiveConfig{
+		Hash: agent.effectiveConfigHash,
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: map[string]*protobufs.AgentConfigFile{
+				"": {Body: []byte(agent.effectiveConfig)},
+			},
+		},
+	}
+}
+
+func (agent *Agent) onRemoteConfig(
+	ctx context.Context,
+	config *protobufs.AgentRemoteConfig,
+) (*protobufs.EffectiveConfig, error) {
+	err := agent.applyRemoteConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return agent.composeEffectiveConfig(), nil
+}
+
+type agentConfigFileItem struct {
+	name string
+	file *protobufs.AgentConfigFile
+}
+
+type agentConfigFileSlice []agentConfigFileItem
+
+func (a agentConfigFileSlice) Less(i, j int) bool {
+	return a[i].name < a[j].name
+}
+
+func (a agentConfigFileSlice) Swap(i, j int) {
+	t := a[i]
+	a[i] = a[j]
+	a[j] = t
+}
+
+func (a agentConfigFileSlice) Len() int {
+	return len(a)
+}
+
+func (agent *Agent) applyRemoteConfig(config *protobufs.AgentRemoteConfig) error {
+	if config == nil {
+		return nil
+	}
+
+	agent.logger.Debugf("Received remote config from server, hash=%x.", config.ConfigHash)
+
+	// Begin with local config. We will later merge received configs on top of it.
+	var k = koanf.New(".")
+	if err := k.Load(rawbytes.Provider([]byte(localConfig)), yaml.Parser()); err != nil {
+		return err
+	}
+
+	orderedConfigs := agentConfigFileSlice{}
+	for name, file := range config.Config.ConfigMap {
+		if name == "" {
+			// skip instance config
+			continue
+		}
+		orderedConfigs = append(orderedConfigs, agentConfigFileItem{
+			name: name,
+			file: file,
+		})
+	}
+
+	// Sort to make sure the order of merging is stable.
+	sort.Sort(orderedConfigs)
+
+	// Append instance config as the last item.
+	instanceConfig := config.Config.ConfigMap[""]
+	if instanceConfig != nil {
+		orderedConfigs = append(orderedConfigs, agentConfigFileItem{
+			name: "",
+			file: instanceConfig,
+		})
+	}
+
+	// Merge received configs.
+	for _, item := range orderedConfigs {
```

Comment:
Should the agent check for content type of each file?

Reply:
Good point, but the server example currently does not set the content type so this won't work right now. We can add content type later on both server and client examples in a future PR.